File: componenthub_mcp/providers/digikey.py
"""DigiKey connector — Product Information API v4 (OAuth2 client credentials).

Capabilities: search, details, pricing, availability, datasheet.
Enable with DIGIKEY_CLIENT_ID / DIGIKEY_CLIENT_SECRET.
"""

import time

import httpx

from .. import config
from ..models import (
    CadAsset,
    Capability,
    ComponentDetails,
    ComponentResult,
    Offer,
    PriceBreak,
    SearchQuery,
)
from .base import Provider, ProviderError

_API = "https://api.digikey.com"


class DigiKeyProvider(Provider):
    name = "digikey"
    display_name = "DigiKey"
    capabilities = frozenset(
        {
            Capability.SEARCH,
            Capability.DETAILS,
            Capability.PRICING,
            Capability.AVAILABILITY,
            Capability.DATASHEET,
        }
    )

    def __init__(self) -> None:
        self._token: str | None = None
        self._token_expiry: float = 0.0

    def is_configured(self) -> bool:
        return bool(config.digikey_client_id() and config.digikey_client_secret())

    def missing_config(self) -> str | None:
        if self.is_configured():
            return None
        return "Set DIGIKEY_CLIENT_ID and DIGIKEY_CLIENT_SECRET (developer.digikey.com)"

    async def _get_token(self, client: httpx.AsyncClient) -> str:
        if self._token and time.monotonic() < self._token_expiry - 60:
            return self._token
        resp = await client.post(
            f"{_API}/v1/oauth2/token",
            data={
                "client_id": config.digikey_client_id(),
                "client_secret": config.digikey_client_secret(),
                "grant_type": "client_credentials",
            },
        )
        if resp.status_code != 200:
            raise ProviderError(f"digikey: token request failed ({resp.status_code})")
        data = resp.json()
        self._token = data["access_token"]
        self._token_expiry = time.monotonic() + int(data.get("expires_in", 600))
        return self._token

    async def _headers(self, client: httpx.AsyncClient) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {await self._get_token(client)}",
            "X-DIGIKEY-Client-Id": config.digikey_client_id() or "",
            "X-DIGIKEY-Locale-Site": "US",
            "X-DIGIKEY-Locale-Currency": "USD",
        }

    def _map_product(self, p: dict) -> ComponentResult:
        variations = p.get("ProductVariations") or [{}]
        best = variations[0]
        breaks = [
            PriceBreak(quantity=b.get("BreakQuantity", 1), unit_price=b.get("UnitPrice", 0.0))
            for b in (best.get("StandardPricing") or [])
        ]
        offer = Offer(
            provider=self.name,
            part_id=best.get("DigiKeyProductNumber") or p.get("ManufacturerProductNumber", ""),
            product_url=p.get("ProductUrl"),
            stock=p.get("QuantityAvailable"),
            price_breaks=breaks,
            packaging=(best.get("PackageType") or {}).get("Name"),
        )
        return ComponentResult(
            mpn=p.get("ManufacturerProductNumber", ""),
            manufacturer=(p.get("Manufacturer") or {}).get("Name"),
            description=(p.get("Description") or {}).get("ProductDescription"),
            category=(p.get("Category") or {}).get("Name"),
            datasheet_url=p.get("DatasheetUrl"),
            image_url=p.get("PhotoUrl"),
            offers=[offer],
        )

    async def search(self, query: SearchQuery) -> list[ComponentResult]:
        async with httpx.AsyncClient(timeout=20) as client:
            body: dict = {"Keywords": query.keyword, "Limit": query.max_results, "Offset": 0}
            if query.manufacturer:
                body["FilterOptionsRequest"] = {
                    "ManufacturerFilter": [{"Id": query.manufacturer}]
                }
            resp = await client.post(
                f"{_API}/products/v4/search/keyword",
                json=body,
                headers=await self._headers(client),
            )
            if resp.status_code != 200:
                raise ProviderError(f"digikey: search failed ({resp.status_code}): {resp.text[:200]}")
            results = [self._map_product(p) for p in resp.json().get("Products", [])]
            if query.in_stock_only:
                results = [r for r in results if any((o.stock or 0) > 0 for o in r.offers)]
            return results

    async def fetch_details(self, part_id: str) -> ComponentDetails:
        async with httpx.AsyncClient(timeout=20) as client:
            resp = await client.get(
                f"{_API}/products/v4/search/{part_id}/productdetails",
                headers=await self._headers(client),
            )
            if resp.status_code != 200:
                raise ProviderError(f"digikey: details failed ({resp.status_code})")
            p = resp.json().get("Product", {})
            base = self._map_product(p)
            specs = {
                (param.get("ParameterText") or ""): (param.get("ValueText") or "")
                for param in p.get("Parameters", [])
            }
            return ComponentDetails(
                **base.model_dump(),
                specifications=specs,
                lifecycle_status=(p.get("ProductStatus") or {}).get("Status"),
                cad_assets=[],
            )

    async def fetch_pricing(self, part_id: str) -> Offer:
        details = await self.fetch_details(part_id)
        if not details.offers:
            raise ProviderError("digikey: no pricing returned")
        return details.offers[0]

    async def fetch_datasheet(self, part_id: str) -> str | None:
        return (await self.fetch_details(part_id)).datasheet_url

    async def fetch_models(self, part_id: str) -> list[CadAsset]:
        raise ProviderError("digikey: CAD models not offered; use snapmagic")

File: lib/domain/providers/digikey.ts
/**
 * DigiKey connector — Product Information API v4 (OAuth2 client credentials).
 * Ports componenthub_mcp/providers/digikey.py exactly.
 *
 * Capabilities: search, details, pricing, availability, datasheet.
 * Enable with DIGIKEY_CLIENT_ID / DIGIKEY_CLIENT_SECRET.
 */
import { digikeyClientId, digikeyClientSecret } from "../config";
import type {
  CadAsset,
  ComponentDetails,
  ComponentResult,
  Offer,
  SearchQuery,
} from "../models";
import { Capability } from "../models";
import { Provider, ProviderError } from "./base";

const API = "https://api.digikey.com";

export class DigiKeyProvider extends Provider {
  readonly name = "digikey";
  readonly displayName = "DigiKey";
  readonly capabilities = new Set([
    Capability.SEARCH,
    Capability.DETAILS,
    Capability.PRICING,
    Capability.AVAILABILITY,
    Capability.DATASHEET,
  ]);

  private _token: string | null = null;
  private _tokenExpiry: number = 0;

  isConfigured(): boolean {
    return Boolean(digikeyClientId() && digikeyClientSecret());
  }

  missingConfig(): string | null {
    if (this.isConfigured()) return null;
    return "Set DIGIKEY_CLIENT_ID and DIGIKEY_CLIENT_SECRET (developer.digikey.com).";
  }

  private async _getToken(): Promise<string> {
    if (this._token && Date.now() / 1000 < this._tokenExpiry - 60) {
      return this._token;
    }
    const clientId = digikeyClientId();
    const clientSecret = digikeyClientSecret();
    const resp = await fetch(`${API}/v1/oauth2/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        client_id: clientId ?? "",
        client_secret: clientSecret ?? "",
        grant_type: "client_credentials",
      }),
    });
    if (!resp.ok) {
      throw new ProviderError(`digikey: token request failed (${resp.status})`);
    }
    const data = (await resp.json()) as {
      access_token: string;
      expires_in?: number;
    };
    this._token = data.access_token;
    this._tokenExpiry = Date.now() / 1000 + (data.expires_in ?? 600);
    return this._token;
  }

  private async _headers(): Promise<Record<string, string>> {
    return {
      Authorization: `Bearer ${await this._getToken()}`,
      "X-DIGIKEY-Client-Id": digikeyClientId() ?? "",
      "X-DIGIKEY-Locale-Site": "US",
      "X-DIGIKEY-Locale-Currency": "USD",
    };
  }

  private _mapProduct(p: Record<string, unknown>): ComponentResult {
    const variations = (p.ProductVariations as Record<string, unknown>[]) ?? [{}];
    const best = variations[0] ?? {};
    const standardPricing = (best.StandardPricing as Record<string, unknown>[]) ?? [];
    const breaks = standardPricing.map((b) => ({
      quantity: (b.BreakQuantity as number) ?? 1,
      unit_price: (b.UnitPrice as number) ?? 0,
      currency: "USD",
    }));
    const packageType = best.PackageType as Record<string, unknown> | undefined;
    const manufacturer = p.Manufacturer as Record<string, unknown> | undefined;
    const description = p.Description as Record<string, unknown> | undefined;
    const category = p.Category as Record<string, unknown> | undefined;
    const offer: Offer = {
      provider: this.name,
      part_id:
        (best.DigiKeyProductNumber as string) ??
        (p.ManufacturerProductNumber as string) ??
        "",
      product_url: (p.ProductUrl as string) ?? null,
      stock: (p.QuantityAvailable as number) ?? null,
      price_breaks: breaks,
      packaging: (packageType?.Name as string) ?? null,
    };
    return {
      mpn: (p.ManufacturerProductNumber as string) ?? "",
      manufacturer: (manufacturer?.Name as string) ?? null,
      description: (description?.ProductDescription as string) ?? null,
      category: (category?.Name as string) ?? null,
      datasheet_url: (p.DatasheetUrl as string) ?? null,
      image_url: (p.PhotoUrl as string) ?? null,
      offers: [offer],
    };
  }

  async search(query: SearchQuery): Promise<ComponentResult[]> {
    const body: Record<string, unknown> = {
      Keywords: query.keyword,
      Limit: query.max_results,
      Offset: 0,
    };
    if (query.manufacturer) {
      body.FilterOptionsRequest = {
        ManufacturerFilter: [{ Id: query.manufacturer }],
      };
    }
    const resp = await fetch(`${API}/products/v4/search/keyword`, {
      method: "POST",
      headers: {
        ...(await this._headers()),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new ProviderError(
        `digikey: search failed (${resp.status}): ${text.slice(0, 200)}`,
      );
    }
    const data = (await resp.json()) as { Products?: Record<string, unknown>[] };
    let results = (data.Products ?? []).map((p) => this._mapProduct(p));
    if (query.in_stock_only) {
      results = results.filter((r) =>
        r.offers.some((o) => (o.stock ?? 0) > 0),
      );
    }
    return results;
  }

  async fetchDetails(partId: string): Promise<ComponentDetails> {
    const resp = await fetch(
      `${API}/products/v4/search/${encodeURIComponent(partId)}/productdetails`,
      {
        method: "GET",
        headers: await this._headers(),
      },
    );
    if (!resp.ok) {
      throw new ProviderError(`digikey: details failed (${resp.status})`);
    }
    const data = (await resp.json()) as { Product: Record<string, unknown> };
    const p = data.Product;
    const base = this._mapProduct(p);
    const parameters = (p.Parameters as Record<string, unknown>[]) ?? [];
    const specs: Record<string, string> = {};
    for (const param of parameters) {
      const key = (param.ParameterText as string) ?? "";
      const val = (param.ValueText as string) ?? "";
      specs[key] = val;
    }
    const productStatus = p.ProductStatus as Record<string, unknown> | undefined;
    return {
      ...base,
      specifications: specs,
      lifecycle_status: (productStatus?.Status as string) ?? null,
      cad_assets: [],
    };
  }

  async fetchPricing(partId: string): Promise<Offer> {
    const details = await this.fetchDetails(partId);
    if (!details.offers.length) {
      throw new ProviderError("digikey: no pricing returned");
    }
    return details.offers[0];
  }

  async fetchDatasheet(partId: string): Promise<string | null> {
    return (await this.fetchDetails(partId)).datasheet_url ?? null;
  }

  async fetchModels(_partId: string): Promise<CadAsset[]> {
    throw new ProviderError("digikey: CAD models not offered; use snapmagic");
  }
}
